Allow redrawing values for clamped nodes when necessary

davidcerny opened the pull request:
This PR is intended to fix issue #767 by making sure that posterior predictive simulations can run (and determine what they should simulate by looking for clamped nodes in the model) while still throwing an error for code such as the following:

```r
x ~ dnNorm(1, 1)
x.clamp(1)
x.redraw()
```

I was surprised to find out that my [original idea](https://github.com/revbayes/revbayes/issues/767#issuecomment-2923845795):

> maybe we could add a boolean member variable to `StochasticNode`, check for it in `redraw()`, let `clamp()` set it to `false`, but provide a dedicated setter that could override it and that we could employ in `PosteriorPredictiveSimulation::run()`

was already mostly implemented: `StochasticNode` has a protected member variable called `ignore_redraw`, which is set to `false` by default and can be toggled by a setter called `.setIgnoreRedraw()`. However, we don't employ this setter anywhere except when implementing the `.redraw()` method for (continuous) stochastic nodes, where it is always set to `false` before a redraw is attempted.

Here, I have repurposed the `ignore_redraw` variable to make sure that: (1) it is always set to `true` by `.clamp()` and to `false` by `.unclamp()`; (2) it is _not_ automatically set to `false` when the user tries to call `.redraw()` directly from Rev; and (3) `PowerPosteriorSimulation` sets it to `false` for any clamped node it finds.

Note that the following still works:

```r
x ~ dnNorm(1, 1)
x.setValue(1)
x.redraw()
```

bredelings commented:
I kind of don't like this approach.  I don't think it is a good idea to introduce (or use) additional state variables like "allow redraw".  This goes from 2 states (clamped/unclamped) to 4 states (clamped-redraw, clamped-noredraw, unclamped-redraw,unclamped-noredraw).

I think we should instead allow making a copy of the model with all variables unclamped.

davidcerny commented:
> I think we should instead allow making a copy of the model with all variables unclamped.

As pointed out in #767, that won't work. The posterior predictive simulation wouldn't know what to simulate.

bredelings commented:
Good point.  The simple alternative would be to simulate every that isn't in the trace.  I left some wider thoughts on #767.

davidcerny commented:
I'm going to close this PR, since I agree it's not a very elegant solution to the problem. But your comments make me wonder if we should go further and get rid of the `ignore_redraw` variable altogether @bredelings. We check for its value in `StochasticNode::redraw()`, but when do we actually expect the check to fail? Currently (i.e., sans the changes proposed in this PR), the only operation that sets it to `true` is `.setValue()`. But `.setValue()` doesn't prevent a node's value from being moved during the MCMC (that would be counterproductive, since we mostly use it to set starting values), nor does it prevent us from redrawing it outside the MCMC (since a call to `.redraw()` resets `ignore_redraw` to `false`).

For what it's worth, `ignore_redraw` was originally added in [this commit](https://github.com/revbayes/revbayes.archive/commit/cfe1e36adbc422a3cdeb42275dd3daf0b9ff275d) by @mlandis, but I certainly wouldn't expect him to remember the exact reason for doing so 12 years later! 😛

davidcerny commented:
@bredelings So no, removing `ignore_redraw` breaks the coalescent tests. My guess is that it's because these make heavy use of `.setValue()`, which doesn't work properly without `ignore_redraw`. It's interesting, though, because there are other tests that include `.setValue()` (e.g., [`test_DEC`](https://github.com/revbayes/revbayes/blob/master/tests/test_DEC/scripts/mcmc_DEC.Rev), [`test_HSMRF_BDP`](https://github.com/revbayes/revbayes/blob/master/tests/test_HSMRF_BDP/scripts/mcmc_HSMRF_BDP.Rev), [`test_deterministic_index`](https://github.com/revbayes/revbayes/blob/master/tests/test_deterministic_index/scripts/test.Rev)), and these don't fail when `ignore_redraw` is removed. So we're either lucky that we just recently added the coalescent tests (because without them, the subtle changes in the behavior of `.setValue()` would go undetected), or the coalescent code is unusually fragile and potentially problematic...

Given the difference between `.setValue()` and `.clamp()`, I would say that we essentially already have 4 different states: fixed-redrawable, fixed-nonredrawable, random-redrawable, and (in theory, I guess) random-nonredrawable.
